xayn-ai/src/ranker/mod.rs

```diff
@@ -43,4 +64,143 @@ impl Ranker {
     pub(crate) fn compute_smbert(&self, sequence: &str) -> Result<Embedding, Error> {
         self.smbert.run(sequence).map_err(Into::into)
     }
+
+    /// Ranks the given documents based on the learned user interests.
+    ///
+    /// # Errors
+    ///
+    /// Fails if no user interests are known.
+    pub(crate) fn rank(&self, documents: &mut [Document]) -> Result<(), Error> {
+        rank(
+            documents,
+            &self.user_interests,
+            self.coi_config.neighbors.get(),
+        )
+    }
+}
+
+fn rank(
+    documents: &mut [Document],
+    user_interests: &UserInterests,
+    neighbors: usize,
+) -> Result<(), Error> {
+    let cois_for_docs = compute_cois_for_docs(documents, user_interests, neighbors)?;
+    let context_for_docs = compute_score_for_docs(cois_for_docs.as_slice());
+
+    documents.sort_unstable_by(|a, b| {
+        nan_safe_f32_cmp(
+            context_for_docs.get(&b.id).unwrap(),
+            context_for_docs.get(&a.id).unwrap(),
+        )
+    });
+
+    Ok(())
+}
+
+fn compute_cois_for_docs(
+    documents: &[Document],
+    user_interests: &UserInterests,
+    neighbors: usize,
+) -> Result<Vec<(DocumentId, CoiComponent)>, Error> {
+    documents
+        .iter()
+        .map(|document| {
+            let coi =
+                compute_coi_for_embedding(&document.smbert_embedding, user_interests, neighbors)
+                    .ok_or(RankerError::NoUserInterests)?;
+            Ok::<(DocumentId, CoiComponent), Error>((document.id, coi))
+        })
+        .collect()
+}
+
+fn compute_score_for_docs(
+    cois_for_docs: &[(DocumentId, CoiComponent)],
+) -> HashMap<&DocumentId, f32> {
+    let context = Context::from_cois(cois_for_docs);
+    let mut scores = HashMap::new();
+    cois_for_docs.iter().for_each(|(id, coi)| {
+        scores.insert(
+            id,
+            context.calculate_score(coi.pos_distance, coi.neg_distance),
+        );
+    });
+    scores
```

Comment:
this could be collected directly:
```suggestion
    cois_for_docs
        .iter()
        .map(|(id, coi)| {
            (
                id,
                context.calculate_score(coi.pos_distance, coi.neg_distance),
            )
        })
        .collect()
```

Reply:
nice!